# Implementation of Supervised Learning Algorithms in Breast Cancer dataset

### Objective:
The objective of this assessment is to evaluate your understanding and ability to apply supervised learning techniques to a real-world dataset.

### Loading and Preprocessing

In [118]:
# Importing necesory libraries 
import numpy as np
import pandas as pd
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns

In [119]:
from sklearn.datasets import load_breast_cancer

In [120]:
data = load_breast_cancer(as_frame = True)

In [121]:
df = data.frame

In [122]:
df.head() # to show the first 5 rows of the dataset

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

In [123]:
df.tail() # To show the last 5 rows of the dataset

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  target  
564                0.2216          0.2060                  0.07115       0  
565                0.1628          0.2572                  0.06637       0  
566                0.1418          0.2218                  0.07820       0  
567                0.2650          0.4087                  0.12400       0  
568                0.0000          0.2871                  0.07039       1  

[5 rows x 31 columns]

In [124]:
df.columns  # To show the column headings

Index(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error', 'fractal dimension error',
       'worst radius', 'worst texture', 'worst perimeter', 'worst area',
       'worst smoothness', 'worst compactness', 'worst concavity',
       'worst concave points', 'worst symmetry', 'worst fractal dimension',
       'target'],
      dtype='object')

In [125]:
df.info()  # to show the basic info of the datset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [126]:
df.shape  # to show the number of rows and columns of dataset

(569, 31)

In [127]:
df.size  # to show the size of the dataset

17639

In [128]:
df.describe()  # to show the statiscal analysis of the dataset

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension      target  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

In [129]:
df.isnull().sum()  # To show the null values in the dataset

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
target                     0
dtype: int64

In [130]:
df.duplicated().sum()  # to show if any duplicates in the dataset

0

In [131]:
df['target']=df.target  # target column assigned 

In [132]:
print("\n Target Value Count")
print(df['target'].value_counts(normalize=True))   


 Target Value Count
target
1    0.627417
0    0.372583
Name: proportion, dtype: float64


### Feature Scaling

In [133]:
from sklearn.preprocessing import StandardScaler


# Create a scaler object
scaler = StandardScaler()

df_scaled = scaler.fit_transform(df)   # scaling the dataset to increase the performance



# Print the scaled data
print("DF Scaled\n")
print(df_scaled.round(2))


DF Scaled

[[ 1.1  -2.07  1.27 ...  2.75  1.94 -1.3 ]
 [ 1.83 -0.35  1.69 ... -0.24  0.28 -1.3 ]
 [ 1.58  0.46  1.57 ...  1.15  0.2  -1.3 ]
 ...
 [ 0.7   2.05  0.67 ... -1.1  -0.32 -1.3 ]
 [ 1.84  2.34  1.98 ...  1.92  2.22 -1.3 ]
 [-1.81  1.22 -1.81 ... -0.05 -0.75  0.77]]


### Explanation:
The breast cancer dataset contains 569 samples with 30 features, and 2 classes (malignant or benign). There are no missing values in this dataset. However, to demonstrate preprocessing steps. Then, I applied StandardScaler to scale the features to have zero mean and unit variance, which is a common requirement for many machine learning algorithms.Rounds the scaled values to 2 decimal places to display data for better readabitiy.This preprocessing step ensures that the features are standardized to have a mean of zero and a standard deviation of one. Such standardization is a common requirement for many machine learning algorithms, as it helps them perform more effectively


### Train_Test_Split 

In [134]:
# Importing train_test_split to split the dataset 
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(df_scaled,df.target,test_size=0.2,random_state=0)

### Classification Algorithm Implementation

In [135]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier


In [136]:
# Initialize classifiers
logreg = LogisticRegression(random_state=42)
dtree = DecisionTreeClassifier(random_state=42)
rf = RandomForestClassifier(random_state=42)
svm = SVC(random_state=42)
knn = KNeighborsClassifier()

In [137]:
# Fit models to the scaled data using LogisticRegression
logreg.fit(x_train, y_train)


LogisticRegression(random_state=42)

In [138]:
# Fit models to the scaled data using DecisionTreeClassifier
dtree.fit(x_train, y_train)


DecisionTreeClassifier(random_state=42)

In [139]:
# Fit models to the scaled data using RandomForestClassifier
rf.fit(x_train, y_train)


RandomForestClassifier(random_state=42)

In [140]:
# Fit models to the scaled data using SVC
svm.fit(x_train, y_train)


SVC(random_state=42)

In [141]:
# Fit models to the scaled data using KNeighborsClassifier
knn.fit(x_train, y_train)

KNeighborsClassifier()

In [142]:
# Predictions on test set of LogisticRegression model
y_pred_logreg = logreg.predict(x_test)

In [143]:
# Predictions on test set of DecisionTreeClassifier model
y_pred_dtree = dtree.predict(x_test)

In [144]:
# Predictions on test set of RandomForestClassifier model
y_pred_rf = rf.predict(x_test)

In [145]:
# Predictions on test set of SVC model
y_pred_svm = svm.predict(x_test)

In [146]:
# Predictions on test set of KNeighborsClassifier model
y_pred_knn = knn.predict(x_test)


### Brief description and justification for each algorithm:

1. Logistic Regression: A linear model that predicts the probability of the target class. Suitable for this dataset since it's a binary classification problem.
2. Decision Tree Classifier: A tree-based model that splits the data into subsets based on features. Suitable since it can handle non-linear relationships and is interpretable.
3. Random Forest Classifier: An ensemble model that combines multiple decision trees. Suitable since it reduces overfitting and improves generalization.
4. Support Vector Machine (SVM): A kernel-based model that finds the optimal hyperplane separating classes. Suitable since it can handle high-dimensional data and is robust to noise.
5. k-Nearest Neighbors (k-NN): A distance-based model that predicts the target class based on nearest neighbors. Suitable since it's simple and can handle non-linear relationships.


### Model Comparison

In [152]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Initialize lists to store the results
results = []

# Evaluate performance
models = [logreg, dtree, rf, svm, knn]
model_names = ["Logistic Regression", "Decision Tree", "Random Forest", "SVM", "k-NN"]

for model, name in zip(models, model_names):
    acc = accuracy_score(y_test, model.predict(x_test))
    prec = precision_score(y_test, model.predict(x_test))
    rec = recall_score(y_test, model.predict(x_test))
    f1 = f1_score(y_test, model.predict(x_test))
    results.append((name, acc, prec, rec, f1))
    print(f"{name}: Accuracy={acc:.4f}, Precision={prec:.4f}, Recall={rec:.4f}, F1-score={f1:.4f}")
    


Logistic Regression: Accuracy=1.0000, Precision=1.0000, Recall=1.0000, F1-score=1.0000
Decision Tree: Accuracy=1.0000, Precision=1.0000, Recall=1.0000, F1-score=1.0000
Random Forest: Accuracy=1.0000, Precision=1.0000, Recall=1.0000, F1-score=1.0000
SVM: Accuracy=1.0000, Precision=1.0000, Recall=1.0000, F1-score=1.0000
k-NN: Accuracy=0.9912, Precision=0.9853, Recall=1.0000, F1-score=0.9926


In [148]:
from sklearn.metrics import classification_report

# Evaluate performance for each model
for model, name in zip(models, model_names):
    y_pred = model.predict(x_test)
    report = classification_report(y_test, y_pred, target_names=["Benign", "Malignant"])
    print(f"{name}:\n{report}\n")


Logistic Regression:
              precision    recall  f1-score   support

      Benign       1.00      1.00      1.00        47
   Malignant       1.00      1.00      1.00        67

    accuracy                           1.00       114
   macro avg       1.00      1.00      1.00       114
weighted avg       1.00      1.00      1.00       114


Decision Tree:
              precision    recall  f1-score   support

      Benign       1.00      1.00      1.00        47
   Malignant       1.00      1.00      1.00        67

    accuracy                           1.00       114
   macro avg       1.00      1.00      1.00       114
weighted avg       1.00      1.00      1.00       114


Random Forest:
              precision    recall  f1-score   support

      Benign       1.00      1.00      1.00        47
   Malignant       1.00      1.00      1.00        67

    accuracy                           1.00       114
   macro avg       1.00      1.00      1.00       114
weighted avg       1.

In [149]:
from sklearn.model_selection import cross_val_score

# Initialize classifiers (same as before)
models = [logreg, dtree, rf, svm, knn]

# Evaluate performance using cross-validation
for model, name in zip(models, model_names):
    scores = cross_val_score(model, x_train, y_train, cv=5, scoring='accuracy')
    avg_accuracy = np.mean(scores)
    print(f"{name}: Average Accuracy = {avg_accuracy:.4f}")


Logistic Regression: Average Accuracy = 0.9978
Decision Tree: Average Accuracy = 1.0000
Random Forest: Average Accuracy = 1.0000
SVM: Average Accuracy = 0.9956
k-NN: Average Accuracy = 0.9978


In [154]:
results_df = pd.DataFrame(results, columns=["Model", "Accuracy", "Precision", "Recall", "F1-score"])

# Find the best and worst models based on F1-score
best_model = results_df.loc[results_df["F1-score"].idxmax()]
worst_model = results_df.loc[results_df["F1-score"].idxmin()]

print(f"\nBest Model: {best_model['Model']} with F1-score={best_model['F1-score']:.4f}")
print(f"Worst Model: {worst_model['Model']} with F1-score={worst_model['F1-score']:.4f}")


Best Model: Logistic Regression with F1-score=1.0000
Worst Model: k-NN with F1-score=0.9926


While all models performed excellently, Logistic Regression along with Decision Tree, Random Forest and SVM stands out as the best due to its perfect scores across all metrics. On the other hand, k-NN, despite being the worst, still shows strong performance with a high F1-score, indicating that it is also a viable option but slightly less optimal compared to the others.
